product-classification/product_classification/InferenceEngine.py

```diff
@@ -0,0 +1,254 @@
+            categories and corresponding thresholds.
+        """
+
+        # add fallback threshold if a shop/category was not present during evaluation
+        # use smallest threshold from all shops as fallback
+        fallback_thresh_by_category = (
+            self.shop_thresholds.groupby("predicted_category")["threshold"].agg(min).to_dict()
+        )
+        fallback_thresholds = classification_df["predicted_category"].apply(
+            lambda x: fallback_thresh_by_category.get(x)
+        )
+
+        # set shop specific thresholds if source and merchant are sent along with request
+        if product_df is not None and {"source", "merchant"}.issubset(set(product_df.columns)):
+            combined = self.join_features_with_inference(product_df, classification_df)
+            join_keys = ["ml_model_name", "source", "merchant", "predicted_category"]
+            combined = combined.join(self.shop_thresholds.set_index(join_keys), on=join_keys)
+            combined["threshold"].fillna(fallback_thresholds)
+        else:
+            combined = pd.concat(
+                [classification_df, fallback_thresholds.rename("threshold")], axis=1
+            )
+
+        combined["category_thresholded"] = combined.apply(
+            lambda x: x["predicted_category"]
+            if x["confidence"] >= x["threshold"]
+            else "under_threshold",
+            axis=1,
+        )
+
+        return combined[list(classification_df.columns) + ["category_thresholded", "threshold"]]
+
+    def probas_to_ProductClassifications(self, probas: pd.DataFrame) -> pd.DataFrame:
+        """
+        This function is used to transform predicted probabilities into a DataFrame with complete
+        ProductClassification objects.
+
+        Args:
+            probas (pd.Dataframe): Dataframe with predicted probabilities.
+
+        Returns:
+            pd.DataFrame: pd.DataFrame of ProductClassification objects.
+        """
+
+        probas.columns = le.inverse_transform(probas.columns)
+        predicted_category = [probas.columns[np.argmax(p)] for p in probas.values]
+        confidence = [np.max(p) for p in probas.values]
+
+        result = pd.DataFrame(
+            {
+                "id": probas.index,
+                "ml_model_name": self.name,
+                "predicted_category": predicted_category,
+                "confidence": confidence,
+                "all_predicted_probabilities": probas.to_dict(orient="records"),
+            }
+        )
+
+        return result
+
+    def run_pipeline(self, request_data: str, apply_thresholds: bool = False) -> pd.DataFrame:
+        """The standard pipeline to run for evaluating the request data and performing inference.
+
+        Args:
+        request_data (json): data which was sent along the POST request. Should include a
+        pd.DataFrame that was transformed into serialized json format.
+        data_format (str): format to use for evaluating the request data. Either 'products' or
```

Comment:
Missing the apply_threshold in the docs, although it seems that should apply shop thresholds, and not thresholds in general?